### PR TITLE
52 watchdog dagger

### PR DESCRIPTION
**Pull request by Benjamin-A**
checks if threads are deadlocked and kills the program if that happens

### CODE REVIEW

File: The-ELEVATOR/elevDriver/elevDriver.go [hunk 1]
```diff
@@ -30,7 +31,7 @@ func InitElevDriver(
 	logger.Info("Driver", "Homing down to nearest floor")
 	elevio.SetMotorDirection(state.Direction)
 
-	go elevatorPollThread(changedState, arrivedAtFloor, stateUpdated)
+	go elevatorPollThread(elevdriverthreadHealthUpdate, changedState, arrivedAtFloor, stateUpdated)
 	go elevatorControlThread(newOrder, arrivedAtFloor, stateUpdated, changedState)
```
Comment (Kytzis) [inline The-ELEVATOR/elevDriver/elevDriver.go:35]: shouldnt they both have an health update?

File: The-ELEVATOR/elevDriver/elevDriver.go [hunk 2]
```diff
@@ -129,7 +131,13 @@ func elevatorPollThread(
 			elevio.SetMotorDirection(state.Direction)
 			logger.Info("Driver", fmt.Sprintf("Door closed. Responding to order %d %d", state.Order.Floor, state.Order.Type))
 			changedState <- state
+		case <-time.After(time.Millisecond * 500):
+			// Continue if no button event is received within 50 milliseconds
+			print("No elevDriver event received for 500 milliseconds")
```
Comment (Kytzis) [inline The-ELEVATOR/elevDriver/elevDriver.go:136]: first of all this is a syntax error, but dont print, just add a continue

File: The-ELEVATOR/buttons/buttons.go [hunk 3]
```diff
@@ -23,6 +25,11 @@ func ButtonThread(
 				Type:  e.Button,
 			}
 			newOrder <- order
+		case <-time.After(time.Millisecond * 500):
+			// Continue if no button event is received within 50 milliseconds
+			print("No button event received for 500 milliseconds")
```
Comment (Kytzis) [inline The-ELEVATOR/buttons/buttons.go:30]: change with continue before merge

File: The-ELEVATOR/network/network.go [hunk 4]
```diff
@@ -10,12 +10,19 @@ var ID int
...
 	newOrder chan structs.NetworkMsg,
 	receivedOrders chan structs.NetworkMsg,
 	commonPort int,
 ) {
 	go networkRxThread(receivedOrders, commonPort)
 	go networkTxThread(newOrder, commonPort)
+
+	for {
+		//tell the watchdog thread that we are alive
+		networkThreadHealthUpdate <- 1
```
Comment (Kytzis) [inline The-ELEVATOR/network/network.go:23]: Id rather see this happen inside the Rx and Tx threads that actually do something and can lockup

File: The-ELEVATOR/network/network.go [hunk 5]
```diff
@@ -47,6 +54,11 @@ func networkTxThread(newOrder chan structs.NetworkMsg, commonPort int) {
 				txCh <- ourState
 			}
 			time.Sleep(txDelay)
+
+		case <-time.After(time.Millisecond * 500):
+			// Continue if no button event is received within 50 milliseconds
+			print("No network event received for 500 milliseconds")
```
Comment (Kytzis) [inline The-ELEVATOR/network/network.go:60]: continue

File: The-ELEVATOR/orders/orders.go [hunk 6]
```diff
@@ -31,6 +33,7 @@ func OrdersThread(
 	for {
 		select {
 		case e := <-newOrder:
+			fmt.Println("New Order")
```
Comment (Kytzis) [inline The-ELEVATOR/orders/orders.go:36]: logger

File: The-ELEVATOR/orders/orders.go [hunk 7]
```diff
@@ -67,6 +70,7 @@ func OrdersThread(
 			}
 
 		case e := <-newDriverState:
+			fmt.Println("New Driver State")
```
Comment (Kytzis) [inline The-ELEVATOR/orders/orders.go:73]: logger

File: The-ELEVATOR/orders/orders.go [hunk 8]
```diff
@@ -104,13 +108,19 @@ func OrdersThread(
...
 				OrderMsg: matrix,
 			}
 
 		case <-rxOrders:
-			continue //fmt.Printf("%+v\n", e)
+			break //fmt.Printf("%+v\n", e)
+
+		case <-time.After(time.Millisecond * 500):
+			// Continue if no button event is received within 50 milliseconds
+			print("No order event received for 500 milliseconds")
```
Comment (Kytzis) [inline The-ELEVATOR/orders/orders.go:120]: continue

File: The-ELEVATOR/watchdog/watchdog.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+			timeOfSuccessfullThreadExecutionMap["NetworkThread"] = time.Now()
+			countNetwork++
+
+		}
+		//check if all threads are responding in time
+		for thread, threadTime := range timeOfSuccessfullThreadExecutionMap {
+
+			if time.Since(threadTime) > time.Duration(WatchdogThreadTimeThresholdInNanoSeconds) {
+				logger.Error("MAIN", fmt.Sprintf("%+v died\n", thread))
+				logger.Error("MAIN", "Killing this primary process pair\n")
```
Comment (Kytzis) [inline The-ELEVATOR/watchdog/watchdog.go:59]: "Committing Seppuku" isntead?

File: The-ELEVATOR/watchdog/watchdog.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+				//kill the program and let the secondary process pair take over
+				os.Exit(1)
+			}
+
+		}
+		logger.Info("WATCHDOG", fmt.Sprintf("Main: %d Buttons: %d Orders: %d ElevDriver: %d Network: %d", countMain, countButtons, countOrders, countElevDriver, countNetwork))
```
Comment (Kytzis) [inline The-ELEVATOR/watchdog/watchdog.go:65]: Make sure to add WATCHDOG as a thread in logger, maybe also use PascalCase as we do with the other threads in logger (excpt main >:( )

**Review by Kytzis — CHANGES_REQUESTED** [review]

File: The-ELEVATOR/elevDriver/elevDriver.go (hunk 2 above)
Comment (Benjamin-A) [reply The-ELEVATOR/elevDriver/elevDriver.go:136]: strangely enough, it works 🤔. Will replace with logger

File: The-ELEVATOR/network/network.go (hunk 4 above)
Comment (Benjamin-A) [reply The-ELEVATOR/network/network.go:23]: yes

**Comment by Kytzis** [comment]
Also make sure to do some thorough testing before merging as always

**Review by Kytzis — APPROVED** [review]
lgtm

**Review by Kytzis — APPROVED** [review]

**Pull request merged**